lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +41,28 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, template:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      @name = template.fetch("metadata", {})["name"]
```

Comment:
The user will definitely know, because the template is not valid if this is missing so the deploy can't possibly succeed. Another option is to raise a FatalDeploymentError right here to immediately abort the deploy if it is missing.